### PR TITLE
Automated testing with Mocha

### PR DESCRIPTION
**Pull request by EhsanShahbazii**
/done

### CODE REVIEW

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 1]
```diff
@@ -1,147 +1,147 @@
...
+4. المان html مقابل: `<div id="mocha">` برای نمایش نتیجه, Mocha از این تگ استفاده میکند.
+5. با کامند(دستور) `mocha.run()` تست ما شروع میشود.
 
-The result:
+نتیجه:
 
 [iframe height=250 src="pow-1" border=1 edit]
 
-As of now, the test fails, there's an error. That's logical: we have an empty function code in `pow`, so `pow(2,3)` returns `undefined` instead of `8`.
+در حال حاضر، تست ناموفق است و یک خطا وجود دارد. و این خطا منطقی است زیرا: کدی داخل تابع `pow` ننوشته ایم و خالی می باشد، بنابراین `pow(2،3)` به جای `8` `undefined` را برمی گرداند.
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:111]:
```suggestion
در حال حاضر، تست ناموفق است و یک خطا وجود دارد. و این خطا منطقی است زیرا: کدی داخل تابع `pow` ننوشته ایم و خالی می باشد، بنابراین `pow(2,3)` به جای `8` `undefined` را برمی گرداند.
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 2]
```diff
@@ -1,147 +1,147 @@
...
 
-For the future, let's note that there are more high-level test-runners, like [karma](https://karma-runner.github.io/) and others, that make it easy to autorun many different tests.
+برای آینده، بیایید توجه داشته باشیم که تست ککنده های سطح بالای بیشتری مانند [karma](https://karma-runner.github.io/) و دیگران وجود دارند که اجرای خودکار بسیاری از تست‌های مختلف را راحت کرده اند.
 
-## Initial implementation
+## اجرای اولیه(Initial implementation)
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:115]:
```suggestion
## پیاده سازی اولیه(Initial implementation)
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 3]
```diff
@@ -1,147 +1,147 @@
...
-Wow, now it works!
+بسیار عالی! الان کار میکند.
 
 [iframe height=250 src="pow-min" border=1 edit]
 
-## Improving the spec
+## بهبود مشخصات(spec)
 
-What we've done is definitely a cheat. The function does not work: an attempt to calculate `pow(3,4)` would give an incorrect result, but tests pass.
+کاری که ما الان انجام دادیم قطعا یک تقلب است. این تابع کار نمی کند: تلاش برای محاسبه `pow(3،4)` نتیجه نادرستی می دهد، اما تست ها با موفقیت انجام می شوند.
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:131]:
```suggestion
کاری که ما الان انجام دادیم قطعا یک تقلب است. این تابع کار نمی کند: تلاش برای محاسبه `pow(3, 4)` نتیجه نادرستی می دهد، اما تست ها با موفقیت انجام می شوند.
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 4]
```diff
@@ -1,147 +1,147 @@
...
 
-...But the situation is quite typical, it happens in practice. Tests pass, but the function works wrong. Our spec is imperfect. We need to add more use cases to it.
+...اما این وضعیت کاملاً عادی است. در عمل اتفاق می افتد و تست ها قبول می شوند، اما عملکرد تابع کاملا اشتباه می باشد. spec(مشخصات) ما ناقص است و ما باید کیس های  بیشتری را به آن اضافه کنیم.
 
-Let's add one more test to check that `pow(3, 4) = 81`.
+بیایید یک تست دیگر اضافه کنیم تا `pow(3، 4) = 81` را بررسی کند.
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:135]:
```suggestion
بیایید یک تست دیگر اضافه کنیم تا `pow(3, 4) = 81` را بررسی کند.
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 5]
```diff
@@ -150,43 +150,43 @@ We can select one of two ways to organize the test here:
...
       });
 
     });
     ```
 
-The principal difference is that when `assert` triggers an error, the `it` block immediately terminates. So, in the first variant if the first `assert` fails, then we'll never see the result of the second `assert`.
+تفاوت اصلی این است که وقتی `assert` باعث ایجاد یک خطا میشود، بلوک `it` بلافاصله پایان می‌یابد. بنابراین، در نوع اول، اگر `assert` اول ناموفق باشد، هرگز نتیجه `assert` دوم را نخواهیم دید.
 
-Making tests separate is useful to get more information about what's going on, so the second variant is better.
+جدا کردن تست‌هااز هم برای دریافت اطلاعات بیشتر در مورد آنچه که اتفاق می‌افتد مفید می باشد، بنابراین نوع دوم بهتر است.
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:171]:
```suggestion
جدا کردن تست‌ها از هم برای دریافت اطلاعات بیشتر در مورد آنچه که اتفاق می‌افتد مفید می باشد، بنابراین نوع دوم بهتر است.
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 6]
```diff
@@ -150,43 +150,43 @@ We can select one of two ways to organize the test here:
...
-The result:
+و نتیجه:
 
 [iframe height=250 src="pow-2" edit border="1"]
 
-As we could expect, the second test failed. Sure, our function always returns `8`, while the `assert` expects `81`.
+همانطور که می‌توانستیم انتظار داشته باشیم، آزمایش دوم رد شد. مطمئناً، تابع ما همیشه `8` را برمی گرداند، در حالی که `assert` انتظار `81` را دارد.
 
-## Improving the implementation
+## بهبود اجرا(Improving the implementation)
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:187]:
```suggestion
## بهبود پیاده سازی(Improving the implementation)
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 7]
```diff
@@ -251,76 +251,76 @@ describe("pow", function() {
   });
 */!*
 
-  // ... more tests to follow here, both describe and it can be added
+  // ... تست‌های بیشتری در اینجا نوشته میشوند، هم توضیح می‌دهند و هم می‌توان آن را اضافه کرد
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:254]:
```suggestion
  // ... را می توان اضافه کرد it و هم descibe تست‌های بیشتری در اینجا نوشته میشوند، هم
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 8]
```diff
@@ -251,76 +251,76 @@ describe("pow", function() {
...
 });
 ```
 
-The nested `describe` defines a new "subgroup" of tests. In the output we can see the titled indentation:
+میبینیم `assert` تودرتو، "زیرگروه(subgroup)" جدیدی از تست‌ها را تعریف می‌کند. در خروجی نیز می توانیم تورفتگی با عنوان را ببینیم:
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:258]:
```suggestion
`describe` تودرتو، "زیرگروه(subgroup)" جدیدی از تست‌ها را تعریف می‌کند. در خروجی نیز می توانیم تورفتگی عنوان تست (title) را ببینیم:
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 9]
```diff
@@ -251,76 +251,76 @@ describe("pow", function() {
...
 
 [iframe height=250 src="pow-4" edit border="1"]
 
-In the future we can add more `it` and `describe` on the top level with helper functions of their own, they won't see `makeTest`.
+در آینده می‌توانیم `it` و `describe` بیشتری در سطح های بالا تری  با توابع کمکی خود اضافه کنیم، آنها `makeTest` را نخواهند دید.
 
-````smart header="`before/after` and `beforeEach/afterEach`"
-We can setup `before/after` functions that execute before/after running tests, and also `beforeEach/afterEach` functions that execute before/after *every* `it`.
+````استفاده هوشمندانه از header="`before/after` و `beforeEach/afterEach`"
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:264]:
```suggestion
````smart header="`before/after` و `beforeEach/afterEach`"
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 10]
```diff
@@ -251,76 +251,76 @@ describe("pow", function() {
...
 ```
 
 [edit src="beforeafter" title="Open the example in the sandbox."]
 
-Usually, `beforeEach/afterEach` and `before/after` are used to perform initialization, zero out counters or do something else between the tests (or test groups).
+معمولاً `beforeEach/afterEach` و `before/after` برای انجام مقداردهی(initialize) اولیه، صفر کردن شمارنده ها یا انجام کار های دیگری بین تست ها (یا گروه های تست) استفاده می شود.
 ````
 
-## Extending the spec
+## تمدید spec
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:302]:
```suggestion
## گسترش spec
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 11]
```diff
@@ -329,26 +329,27 @@ describe("pow", function() {
...
 [iframe height=530 src="pow-nan" edit border="1"]
 
-The newly added tests fail, because our implementation does not support them. That's how BDD is done: first we write failing tests, and then make an implementation for them.
+تست های جدید اضافه شده با شکست مواجه می شوند، زیرا پیاده سازی ما از آنها پشتیبانی نمی کند. BDD اینگونه انجام می شود: ابتدا تست های شکست خورده را می نویسیم و سپس برای آنها پیاده سازی می کنیم.
 
 ```smart header="Other assertions"
-Please note the assertion `assert.isNaN`: it checks for `NaN`.
+لطفاً به پیغام `assert.isNaN` توجه کنید: `NaN` را بررسی می کند.
 
 There are other assertions in [Chai](https://www.chaijs.com/) as well, for instance:
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:341]: Please translate this line as well.

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 12]
```diff
@@ -329,26 +329,27 @@ describe("pow", function() {
...
-
-- `assert.equal(value1, value2)` -- checks the equality  `value1 == value2`.
-- `assert.strictEqual(value1, value2)` -- checks the strict equality `value1 === value2`.
-- `assert.notEqual`, `assert.notStrictEqual` -- inverse checks to the ones above.
-- `assert.isTrue(value)` -- checks that `value === true`
-- `assert.isFalse(value)` -- checks that `value === false`
-- ...the full list is in the [docs](https://www.chaijs.com/api/assert/)
+assert دیگری نیز در [chai](https://www.chaijs.com/) وجود دارد، به عنوان مثال:
+
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:343]:
```suggestion

assert دیگری نیز در [chai](https://www.chaijs.com/) وجود دارد، به عنوان مثال:
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 13]
```diff
@@ -329,26 +329,27 @@ describe("pow", function() {
...
+- `assert.equal(value1, value2)` -- برابری را بررسی می کند  `value1 == value2`.
+- `assert.strictEqual(value1, value2)` -- برابری دقیق را بررسی می کند `value1 === value2`.
+- `assert.notEqual`, `assert.notStrictEqual` -- بررسی معکوس به موارد بالا.
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:346]:
```suggestion
- `assert.notEqual`, `assert.notStrictEqual` -- بررسی معکوس موارد بالا.
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 14]
```diff
@@ -329,26 +329,27 @@ describe("pow", function() {
...
+- `assert.isTrue(value)` -- این را بررسی می کند `value === true`
+- `assert.isFalse(value)` -- این را بررسی می کند `value === false`
+- ...لیست کامل را میتوانید در اینجا مشاهده کنید [docs](https://www.chaijs.com/api/assert/)
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:349]:
```suggestion
- ... لیست کامل را می توانید در [مستندات](https://www.chaijs.com/api/assert/) مشاهده کنید.
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 15]
```diff
@@ -367,43 +368,44 @@ function pow(x, n) {
...
+به طور طبیعی, اصلاح و بهبود کد های تست شده به صورت خودکار آسان تر است. اما دلیل دیگری نیز وجود دارد.
 
-To write tests, the code should be organized in such a way that every function has a clearly described task, well-defined input and output. That means a good architecture from the beginning.
+برای نوشتن تست‌ها، کد باید به گونه‌ای سازماندهی شود که هر تابع دارای یک وظیفه واضح، ورودی و خروجی کاملاً تعریف شده باشد. این یعنی یک معماری خوب از ابتدا.
 
-In real life that's sometimes not that easy. Sometimes it's difficult to write a spec before the actual code, because it's not yet clear how it should behave. But in general writing tests makes development faster and more stable.
+در زندگی واقعی, گاهی اوقات آنقدرها هم آسان نیست. گاهی اوقات نوشتن یک مشخصات(spec) قبل از کد واقعی دشوار است، زیرا هنوز مشخص نیست که چگونه باید رفتار کند. اما به طور کلی تست, توسعه را سریعتر و پایدارتر می کند.
 
-Later in the tutorial you will meet many tasks with tests baked-in. So you'll see more practical examples.
+بعداً در این دوره آموزشی, با بسیاری از وظایف تست های پخته شده(baked-in) روبرو خواهید شد. بنابراین نمونه های کاربردی بیشتری خواهید دید.
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:409]:
```suggestion
بعداً در این دوره آموزشی, با بسیاری از تمرین ها همراه با تست های فیکس شده(baked-in) روبرو خواهید شد. بنابراین نمونه های کاربردی بیشتری خواهید دید.
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 16]
```diff
@@ -367,43 +368,44 @@ function pow(x, n) {
...
+با این spec ها, می‌توانیم با خیال راحت عملکرد تابع را بهبود بخشیم، تغییر دهیم، حتی از ابتدا بازنویسی کنیم و مطمئن شویم که همچنان درست کار می‌کند.
 
-With the spec, we can safely improve, change, even rewrite the function from scratch and make sure it still works right.
+این به ویژه در پروژه های بزرگ زمانی که یک تابع در بسیاری از مکان ها استفاده می شود مهم است. وقتی چنین تابعی را تغییر می‌دهیم، راهی برای بررسی دستی وجود ندارد که آیا هر مکانی که از آن استفاده می‌کند هنوز درست کار می‌کند یا خیر.
 
-That's especially important in large projects when a function is used in many places. When we change such a function, there's just no way to manually check if every place that uses it still works right.
+بدون تست, مردم دو راه دارند:
 
-Without tests, people have two ways:
+1. برای انجام تغییر که مهم نیست که چه باشد, کاربران ما با اشکال مواجه می شوند، زیرا ما احتمالاً چیزی را به صورت دستی بررسی نمی کنیم.
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:394]:
```suggestion
```
No additional line should be added.

**Review by mahdiHash — CHANGES_REQUESTED** [review]
Please review the changes. Notice the apply-to-whole ones.
You can add the suggestions to a batch and submit a single commit ([docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes)).

**Comment by mahdiHash** [comment]
Please translate the tasks, too.

**Comment by EhsanShahbazii** [comment]
ok. I corrected the file.

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 17]
```diff
@@ -329,26 +329,28 @@ describe("pow", function() {
...
+و نتیجه با تست جدید به صورت زیر می باشد:
 
 [iframe height=530 src="pow-nan" edit border="1"]
 
-The newly added tests fail, because our implementation does not support them. That's how BDD is done: first we write failing tests, and then make an implementation for them.
+تست های جدید اضافه شده با شکست مواجه می شوند، زیرا پیاده سازی ما از آنها پشتیبانی نمی کند. BDD اینگونه انجام می شود: ابتدا تست های شکست خورده را می نویسیم و سپس برای آنها پیاده سازی می کنیم.
 
 ```smart header="Other assertions"
-Please note the assertion `assert.isNaN`: it checks for `NaN`.
+لطفاً به ادعای `assert.isNaN` توجه کنید: `NaN` بودن را بررسی میکند.
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:339]:
```suggestion
لطفاً به `assert.isNaN` توجه کنید: `NaN` بودن را بررسی میکند.
```

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 18]
```diff
@@ -329,26 +329,28 @@ describe("pow", function() {
...
 
 There are other assertions in [Chai](https://www.chaijs.com/) as well, for instance:
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:342]:
```suggestion
```
The lines 341 and 342 must be deleted.

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 19]
```diff
@@ -367,43 +369,44 @@ function pow(x, n) {
...
+با این spec ها, می‌توانیم با خیال راحت عملکرد تابع را بهبود بخشیم، تغییر دهیم، حتی از ابتدا بازنویسی کنیم و مطمئن شویم که همچنان درست کار می‌کند.
 
-With the spec, we can safely improve, change, even rewrite the function from scratch and make sure it still works right.
+این به ویژه در پروژه های بزرگ زمانی که یک تابع در بسیاری از مکان ها استفاده می شود مهم است. وقتی چنین تابعی را تغییر می‌دهیم، راهی برای بررسی دستی وجود ندارد که آیا هر مکانی که از آن استفاده می‌کند هنوز درست کار می‌کند یا خیر.
 
-That's especially important in large projects when a function is used in many places. When we change such a function, there's just no way to manually check if every place that uses it still works right.
+بدون تست, مردم دو راه دارند:
 
-Without tests, people have two ways:
+1. برای انجام تغییر که مهم نیست که چه باشد, کاربران ما با اشکال مواجه می شوند، زیرا ما احتمالاً چیزی را به صورت دستی بررسی نمی کنیم.
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:395]:
```suggestion
```
Like 395 must be deleted.

File: 1-js/03-code-quality/05-testing-mocha/article.md [hunk 20]
```diff
@@ -1,147 +1,147 @@
...
-6. Go to 3, update the implementation till tests give no errors.
-7. Repeat steps 3-6 till the functionality is ready.
+1. یک spec اولیه با تست هایی برای بنیادی(اساسی) ترین عملکرد نوشته شده است.
+2. یک پیاده سازی اولیه ایجاد می شود.
+3. برای بررسی اینکه آیا کار می کند یا نه، فریم ورک تست [Mocha] (https://mochajs.org/) (جزئیات بیشتر به زودی) را اجرا می کنیم که spec را اجرا می کند. تا زمانی که عملکرد کامل نباشد، خطاها نمایش داده می شوند. ما اصلاحات را انجام می دهیم تا زمانی که همه چیز درست کار بکند
+4. اکنون ما یک پیاده سازی اولیه با تست داریم.
+5. کیس های بیشتری را به spec اضافه می کنیم که احتمالاً هنوز توسط پیاده سازی ها پشتیبانی نشده اند. تست ها به مشکل بر میخورند.
+6. به شماره 3 برگردید و پیاده سازی ها را آپدیت کنید تا وقتی که تست ها خطایی ندهند.
+7. مراحل 3-6 را تکرار کنید تا عملکرد ها آماده شود.
+8.
```
Comment (mahdiHash) [inline 1-js/03-code-quality/05-testing-mocha/article.md:77]:
```suggestion

```
Line 77 must be empty.

**Review by mahdiHash — CHANGES_REQUESTED** [review]
Please review the suggestions.
You can add the suggestions to a batch and submit a single commit ([docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes)).

**Comment by mahdiHash** [comment]
> Please translate the tasks, too.

Have you translated the tasks, @EhsanShahbazii?

**Comment by EhsanShahbazii** [comment]
The word task in article? yes

**Comment by mahdiHash** [comment]
> The word task in article? yes @mahdiHash

No, I mean the task [here](https://github.com/EhsanShahbazii/fa.javascript.info/tree/patch-1/1-js/03-code-quality/05-testing-mocha/3-pow-test-wrong).
You translated them in #249, but not in this PR.

**Comment by EhsanShahbazii** [comment]
i see. done @mahdiHash

**Review by mahdiHash — APPROVED** [review]

**Comment by mahdiHash** [comment]
Thanks for your contribution, @EhsanShahbazii.

**Comment by EhsanShahbazii** [comment]
Thank you for your explanation and guidance.🤍

**Pull request merged**